File: core/04-mcp/go/main.go
// main.go —— demo only：让 LLM 通过 Python MCP server 创建并读回一个 todo 文件。
//
// Go 客户端拉起 ../python/server.py 作为子进程；这正是 MCP 的卖点 ——
// 工具实现在哪个语言不重要，client 只看 JSON-RPC 协议。
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"time"

	"github.com/joho/godotenv"
	"github.com/sashabaranov/go-openai"
)

func main() {
	if err := godotenv.Load("../.env"); err != nil {
		log.Println("warning: .env not loaded:", err)
	}

	cfg := openai.DefaultConfig(envOr("API_KEY", "not-needed"))
	cfg.BaseURL = os.Getenv("API_BASE_URL")
	cfg.HTTPClient = &http.Client{
		Transport: &http.Transport{Proxy: nil},
		Timeout:   60 * time.Second,
	}
	c := openai.NewClientWithConfig(cfg)
	model := os.Getenv("MODEL_ID")
	ctx := context.Background()

	workspace, _ := filepath.Abs("test_workspace")
	_ = os.MkdirAll(workspace, 0o755)

	// Python server 在上一层目录
	pyServer, _ := filepath.Abs("../python/server.py")
	mcp, err := NewMCPClient(ctx, "python3", pyServer, workspace)
	if err != nil {
		log.Fatal("mcp start:", err)
	}
	defer mcp.Close()

	bridge, err := NewBridge(mcp)
	if err != nil {
		log.Fatal("bridge:", err)
	}

	for _, q := range []string{
		"请在 todo.txt 里写三条 todo：1. 学习 MCP 2. 写 demo 3. 提 PR",
		"先列出当前目录，然后读 todo.txt 的内容回给我。",
	} {
		fmt.Println(">>>", q)
		ans, err := bridge.Chat(ctx, c, model, q, 6)
		if err != nil {
			fmt.Println("  ERROR:", err)
			continue
		}
		fmt.Println(ans)
		fmt.Println()
	}
}

func envOr(k, fb string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}
	return fb
}

File: core/04-mcp/python/main.py
"""main.py —— demo only：让 LLM 通过 MCP server 创建并读回一个 todo 文件。"""
from __future__ import annotations

import os
import sys

import httpx
from dotenv import load_dotenv
from openai import OpenAI

from client import run_sync

load_dotenv()

_http = httpx.Client(trust_env=False, timeout=60.0)
_client = OpenAI(
    base_url=os.environ["API_BASE_URL"],
    api_key=os.environ.get("API_KEY", "not-needed"),
    http_client=_http,
)
_model = os.environ["MODEL_ID"]

# 用同一个 Python 解释器拉起 server.py；workspace 放当前目录下
_workspace = os.path.abspath("test_workspace")
_server_cmd = sys.executable
_server_args = ["server.py", _workspace]


def main() -> None:
    print(">>> 让 LLM 通过 MCP 创建 todo.txt")
    print(run_sync(_server_cmd, _server_args, _client, _model,
                   "请在 todo.txt 里写三条 todo：1. 学习 MCP 2. 写 demo 3. 提 PR"))

    print("\n>>> 让 LLM 通过 MCP 列目录并读回 todo.txt")
    print(run_sync(_server_cmd, _server_args, _client, _model,
                   "先列出当前目录，然后读 todo.txt 的内容回给我。"))


if __name__ == "__main__":
    main()
